internal/proxy/reverse_proxy.go

```diff
@@ -0,0 +1,161 @@
+
+			dialer := &net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}
+
+			conn, err := dialer.Dial(network, addr)
+			if err != nil {
+				log.Infof(" > RevProxy: Dial - Error: %v", err)
+			}
+			return conn, err
+		},
+
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+
+	return &httputil.ReverseProxy{
+		Director:  director(target),
+		Transport: transport,
+		// ModifyResponse: modifyResponse,
+	}
+}
+
+type responseRecorder struct {
+	http.ResponseWriter
+	statusCode  int
+	err         error
+	wroteHeader bool
+	log         *log.Entry
+}
+
+func newResponseRecorder(w http.ResponseWriter, logger *log.Entry) *responseRecorder {
+	return &responseRecorder{
+		ResponseWriter: w,
+		log:            logger.WithField("response_recorder", "true"),
+		statusCode:     http.StatusOK,
+		err:            nil,
+		wroteHeader:    false,
+	}
+}
+
+func (rr *responseRecorder) WriteHeader(code int) {
+	rr.log.Infof("Write Header called with %d", code)
+
+	if rr.wroteHeader {
+		rr.log.Warnf("Multiple Write Header %d called; SKIPPED", code)
+		return
+	}
+
+	rr.statusCode = code
+	rr.wroteHeader = true
+
+	// refer httputils/reverseproxy.go ServerHTTP which sets the
+	// StatusBadGateway when transport.RoundTrip returns an error
+	if code == http.StatusBadGateway {
+		// set err for bad gateway
+		rr.err = fmt.Errorf("bad gateway error: %d", code)
+		rr.log.Warnf("Error BadGateway WriteHeader %v - SKIPPED", code)
+		return
+	}
+
+	rr.log.Infof(" > RevProxy: WriteHeader %v - OK", code)
+	rr.ResponseWriter.WriteHeader(code)
+}
+
+func (rr *responseRecorder) Write(stuff []byte) (int, error) {
+	if rr.err == nil {
+		rr.log.Infof(" > RevProxy: Write stuff len: %d - OK", len(stuff))
+		return rr.ResponseWriter.Write(stuff)
+	}
+
+	// fake that it wrote it
+	rr.log.Warnf("Faking write len: %d - SKIPPED as err is set", len(stuff))
+	return len(stuff), nil
+}
+
+func (rp *reverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+
+	logger := rp.logger.WithFields(log.Fields{
+		"redirect_url":     rp.redirectURL.String(),
+		"request_url":      req.URL.String(),
+		"response_timeout": rp.responseTimeout,
+	})
+
+	// HACK: do not merge this is just to test timeout
+	timeout := 8 * time.Second
+	if strings.HasSuffix(rp.redirectURL.String(), "blueocean.png") {
+		timeout = time.Duration(rand.Intn(1000)+800) * time.Millisecond
+	}
+
+	// todo: need a good timeout - say 22 sec?
+	// HACK: do not merge this is just to test things out
+	ctx := req.Context()
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	// replace with
+	// ctx, cancel := context.WithTimeout(ctx, rp.responseTimeout)
+	defer cancel()
+
+	outreq := req.WithContext(ctx)
+
+	rr := newResponseRecorder(rw, logger)
+	proxy := newLoggedReverseProxy(req.URL)
+	proxy.ServeHTTP(rr, outreq)
+
+	if rr.err != nil {
+		logger.Warnf("Error %q - code: %d", rr.err, rr.statusCode)
+		http.Redirect(rw, req, rp.redirectURL.String(), http.StatusFound)
```

Comment:
@chmouel this is how user is asked to retry. this returns a 302
see: https://golang.org/pkg/net/http/   StatusFound            = 302 // RFC 7231, 6.4.3

Reply:
It would nice if you tried running this locally and watch the logs; might want to apply 

```patch
diff --git a/cmd/fabric8-jenkins-proxy/main.go b/cmd/fabric8-jenkins-proxy/main.go
index 18bcd6f..628c75b 100644
--- a/cmd/fabric8-jenkins-proxy/main.go
+++ b/cmd/fabric8-jenkins-proxy/main.go
@@ -39,7 +39,8 @@ const (
 var mainLogger = log.WithFields(log.Fields{"component": "main"})
 
 func init() {
-	log.SetFormatter(&log.JSONFormatter{})
+	// log.SetFormatter(&log.JSONFormatter{})
+	log.SetFormatter(&log.TextFormatter{ForceColors: true})
 
 	level := log.InfoLevel
 	switch levelStr, _ := os.LookupEnv("JC_LOG_LEVEL"); levelStr {
```